### PR TITLE
RUM-5004 Record resources while processing snapshots

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Previously, resources where created while recording the views. But in SwiftUI, we will only be able to create resources at the processing phase. To accommodate this, resources will now be created when building the wireframes and written by the snapshot processor.

### How?

The `SnapshotProcessor` now holds a reference to a `ResourceProcessor`, and the `SessionReplayWireframesBuilder` will keep track of resources while building the wireframes. The `SnapshotProcessor` will pass all collected resources to its `ResourceProcessor` instance. 

**Note:**
The `ResourceProcessor` and the `ResourcesWriter` are doing the same checks to deduplicate images, but on different threads. This needs a refactoring as well, but it's outside the scope of this PR.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests for Session Replay
- [ ] Run smoke tests
- [ ] Run tests for `tools/`

### CODE REVIEW

**Review by mariedm — DISMISSED** [review]
Just a few nits & questions, LGMT! 🙌

File: DatadogSessionReplay/Sources/Processor/Builders/WireframesBuilder.swift [hunk 1]
```diff
@@ -223,7 +229,7 @@ extension SessionReplayWireframesBuilder {
         return .webviewWireframe(value: wireframe)
     }
 
-    internal func hiddenWebViewWireframes() -> [SRWireframe] {
+    public func hiddenWebViewWireframes() -> [SRWireframe] {
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogSessionReplay/Sources/Processor/Builders/WireframesBuilder.swift:232]: Just to align with other methods access level

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UIImageViewRecorder.swift [hunk 2]
```diff
@@ -140,26 +133,30 @@ internal struct UIImageViewWireframesBuilder: NodeWireframesBuilder {
...
-                    builder.createImageWireframe(
-                        resourceId: imageResource.calculateIdentifier(),
-                        id: imageWireframeID,
-                        frame: contentFrame,
-                        clip: clipsToBounds ? clip : nil
-                    )
+
+        guard let contentFrame = contentFrame else {
+            return wireframes
+        }
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UIImageViewRecorder.swift:139]: I'm so pre Swift 5.7 😆

File: DatadogSessionReplay/Tests/Processor/SnapshotProcessorTests.swift [hunk 3]
```diff
@@ -281,6 +296,41 @@ class SnapshotProcessorTests: XCTestCase {
         XCTAssertEqual(fullSnapshotRecord.data.wireframes.first?.id, Int64(webview.hash), "The hidden webview wireframe should be first")
     }
 
+    func testWhenProcessingViewTreeSnapshot_itRecordResources() throws {
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogSessionReplay/Tests/Processor/SnapshotProcessorTests.swift]: Yes, let me add the test 👍

**Review by mariedm — APPROVED** [review]
👌

**Pull request merged**